bundles/org.openhab.binding.smarther/src/main/java/org/openhab/binding/smarther/internal/handler/SmartherModuleHandler.java

```diff
@@ -0,0 +1,661 @@
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.smarther.internal.api.dto.Chronothermostat;
+import org.openhab.binding.smarther.internal.api.dto.Enums.BoostTime;
+import org.openhab.binding.smarther.internal.api.dto.Enums.Mode;
+import org.openhab.binding.smarther.internal.api.dto.ModuleStatus;
+import org.openhab.binding.smarther.internal.api.dto.Notification;
+import org.openhab.binding.smarther.internal.api.dto.Program;
+import org.openhab.binding.smarther.internal.api.exception.SmartherGatewayException;
+import org.openhab.binding.smarther.internal.api.exception.SmartherSubscriptionAlreadyExistsException;
+import org.openhab.binding.smarther.internal.config.SmartherModuleConfiguration;
+import org.openhab.binding.smarther.internal.model.ModuleSettings;
+import org.openhab.binding.smarther.internal.util.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@code SmartherModuleHandler} class is responsible of a single Smarther Chronothermostat, handling the commands
+ * that are sent to one of its channels.
+ * Each Smarther Chronothermostat communicates with the Smarther API via its assigned {@code SmartherBridgeHandler}.
+ *
+ * @author Fabio Possieri - Initial contribution
+ */
+@NonNullByDefault
+public class SmartherModuleHandler extends BaseThingHandler {
+
+    private static final String DAILY_MIDNIGHT = "1 0 0 * * ? *";
+    private static final long POLL_INITIAL_DELAY = 5;
+
+    private final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final CronScheduler cronScheduler;
+    private final SmartherDynamicStateDescriptionProvider dynamicStateDescriptionProvider;
+    private final ChannelUID programChannelUID;
+    private final ChannelUID endDateChannelUID;
+
+    // Field members assigned in initialize method
+    private @NonNullByDefault({}) ScheduledCompletableFuture<Void> jobFuture;
+    private @NonNullByDefault({}) Future<?> pollFuture;
+    private @NonNullByDefault({}) SmartherBridgeHandler bridgeHandler;
+    private @NonNullByDefault({}) SmartherModuleConfiguration config;
+
+    // Chronothermostat local status
+    private @NonNullByDefault({}) ExpiringCache<List<Program>> programCache;
+    private @NonNullByDefault({}) Chronothermostat chronothermostat;
+    private @NonNullByDefault({}) ModuleSettings moduleSettings;
```

Comment:
What is the intent and best practice around these?
I can make the `chronothermostat` and `moduleSettings` objects Nullable. The others are actually created in the Initialize method (`jobFuture`, `pollFuture`, `programCache`) and have no reason to be null, or are directly set by the OH platform (`bridgeHandler`, `config`).

Reply:
I have had seen invocations of `handleCommand()` before `initialize()` has been finished. I opened a thread, if this is intended behavior: https://community.openhab.org/t/invocation-of-handlecommand-before-initializing-has-finished/99956

Reply:
Hmmm, it shouldn't be a correct behaviour in my opinion, I'd expect the thing to start accepting commands only after `initialize()` has finished. However, let's see what others will say in the thread 😉

Reply:
Result of the thread is that `handleCommand()` can be invoked after the thing status has been updated to ONLINE, OFFLINE or UNKNOWN. So, you should move `updateStatus(ThingStatus.UNKNOWN)` behind scheduling of the timer tasks.

Reply:
Noted. I've moved it accordingly on `SmartherBridgeHandler`, while on `SmartherModuleHandler` I believe the following to be correct:

> programCache = new ExpiringCache<>(Duration.ofHours(config.getProgramsRefreshPeriod()), this::programCacheAction);
> moduleSettings = new ModuleSettings(config.getPlantId(), config.getModuleId());
> updateStatus(ThingStatus.UNKNOWN);
> scheduleJob();
> schedulePoll(); 

As this way the `handleCommand()` may be called only "after" the variables `programCache` and `moduleSettings` have been initialized, while there's no issue in receiving a command before timed tasks are started with `scheduleJob()` and `schedulePoll()` (these are used to periodically refresh the module in case of no commands).

Reply:
Yes, since you're doing a null check in `dispose()` it should be ok.